xrtl/gfx/util/memory_commands.h

```diff
@@ -0,0 +1,409 @@
+// this means that all data must be POD - a ref_ptr or std::string will not be
+// correctly manipulated!
+//
+// Commands that have variable sizes (such as those accepting data buffers or
+// arrays) are encoded with a count in their command struct and then immediately
+// followed by the encoded variable-length data.
+//
+// The command buffer writes to a slab-style arena allocator with some defined
+// block size (that is a reasonable max of any command + data we'll encode).
+// Each block from the allocator is prefixed with a kBufferPacketFrame command
+// denoting the total size of the packet and a pointer to the next packet.
+
+// Maximum size of a single command in bytes.
+// This is used to size buffers that hold the commands.
+// Note that the 64k here comes from UpdateBuffer's max inline data size.
+constexpr size_t kMaxCommandSize = (4 + 64) * 1024;
+
+enum class CommandType {
+  kBeginTransferCommands = 0,
+  kEndTransferCommands,
+  kBeginComputeCommands,
+  kEndComputeCommands,
+  kBeginRenderCommands,
+  kEndRenderCommands,
+  kBeginRenderPass,
+  kEndRenderPass,
+
+  kSetFence,
+  kResetFence,
+  kWaitFences,
+  kPipelineBarrier,
+  kMemoryBarrier,
+  kBufferBarrier,
+  kImageBarrier,
+  kFillBuffer,
+  kUpdateBuffer,
+  kCopyBuffer,
+  kCopyImage,
+  kCopyBufferToImage,
+  kCopyImageToBuffer,
+  kBlitImage,
+  kResolveImage,
+  kGenerateMipmaps,
+  kClearColorImage,
+  kClearDepthStencilImage,
+  kClearColorAttachment,
+  kClearDepthStencilAttachment,
+  kBindComputePipeline,
+  kBindRenderPipeline,
+  kBindResourceSet,
+  kPushConstants,
+  kDispatch,
+  kDispatchIndirect,
+  kNextSubpass,
+  kSetScissors,
+  kSetViewports,
+  kSetLineWidth,
+  kSetDepthBias,
+  kSetDepthBounds,
+  kSetStencilCompareMask,
+  kSetStencilWriteMask,
+  kSetStencilReference,
+  kSetBlendConstants,
+  kBindVertexBuffers,
+  kBindIndexBuffer,
+  kDraw,
+  kDrawIndexed,
+  kDrawIndirect,
+  kDrawIndexedIndirect,
+};
+
+// Command packet header used by the buffer reader/writer.
+struct PacketHeader {
+  size_t packet_length;       // Excludes this header.
+  PacketHeader* next_packet;  // nullptr if this is the last packet.
+};
+
+// Single command header that prefixes all commands written.
+struct CommandHeader {
+  CommandType command_type;
+};
+
+struct BeginTransferCommandsCommand {};
+struct EndTransferCommandsCommand {};
+struct BeginComputeCommandsCommand {};
+struct EndComputeCommandsCommand {};
+struct BeginRenderCommandsCommand {};
+struct EndRenderCommandsCommand {};
+
+struct BeginRenderPassCommand {
+  RenderPass* render_pass;
+  Framebuffer* framebuffer;
+};
+
+struct EndRenderPassCommand {};
+
+struct SetFenceCommand {
+  CommandFence* fence;
+  PipelineStageFlag pipeline_stage_mask;
+};
+
+struct ResetFenceCommand {
+  CommandFence* fence;
+  PipelineStageFlag pipeline_stage_mask;
+};
+
+struct WaitFencesCommand {
+  size_t fence_count;  // ArrayView<CommandFence*> fences;
```

Comment:
? Not implemented yet but TODO? This would just change to that ArrayView (which has size())?
I see the other ArrayView comments below.